File: MediaLibrary/Classes/MainWindowController.m
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2014 Jean-David Gadina - www-xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#import "MainWindowController.h"
#import "Track.h"

@import iTunesLibrary;

@interface MainWindowController() < NSTableViewDelegate >

@property( atomic, readwrite, strong ) NSArray< Track    * > * allTracks;
@property( atomic, readwrite, strong ) NSArray< Track    * > * tracks;
@property( atomic, readwrite, strong ) NSArray< NSString * > * artists;
@property( atomic, readwrite, strong ) NSArray< NSString * > * albums;

@property( atomic, readwrite, strong ) IBOutlet NSArrayController * tracksController;
@property( atomic, readwrite, strong ) IBOutlet NSArrayController * artistsController;
@property( atomic, readwrite, strong ) IBOutlet NSArrayController * albumsController;

- ( NSArray< NSString * > * )uniqueValue: ( NSString * )keypath inArray: ( NSArray< Track * > * )array;

@end

@implementation MainWindowController

- ( id )init
{
    return [ super initWithWindowNibName: @"MainWindow" ];
}

- ( void )windowDidLoad
{
    NSError   * error;
    ITLibrary * library;
    NSString  * title;
    
    [ super windowDidLoad ];
    
    library = [ ITLibrary libraryWithAPIVersion: @"1.0" error: &error ];
    
    if( library == nil || error != nil )
    {
        NSLog( @"%@", error );
    }
    else
    {
        NSMutableArray< Track * > * tracks = [ NSMutableArray new ];
        
        for( ITLibMediaItem * item in library.allMediaItems )
        {
            [ tracks addObject: [ [ Track alloc ] initWithItem: item ] ];
        }
        
        self.allTracks = [ tracks copy ];
        self.tracks    = self.allTracks;
        self.artists   = [ self uniqueValue: @"artist" inArray: self.tracks ];
        self.albums    = [ self uniqueValue: @"album"  inArray: self.tracks ];
        
        [ self.artistsController setSortDescriptors: @[ [ NSSortDescriptor sortDescriptorWithKey: @"description" ascending: YES ] ] ];
        [ self.albumsController  setSortDescriptors: @[ [ NSSortDescriptor sortDescriptorWithKey: @"description" ascending: YES ] ] ];
        [ self.tracksController  setSortDescriptors:
            @[
                [ NSSortDescriptor sortDescriptorWithKey: @"artist" ascending: YES ],
                [ NSSortDescriptor sortDescriptorWithKey: @"album"  ascending: YES ],
                [ NSSortDescriptor sortDescriptorWithKey: @"title"  ascending: YES ]
            ]
        ];
        
        self.artistsController.selectionIndexes = [ NSIndexSet indexSet ];
        
        title             = [ self.window.title stringByAppendingString: [ NSString stringWithFormat: @" (%lu tracks)", self.allTracks.count ] ];
        self.window.title = title;
    }
}

- ( NSArray< NSString * > * )uniqueValue: ( NSString * )keypath inArray: ( NSArray< ITLibMediaItem * > * )array
{
    NSMutableSet   * set;
    ITLibMediaItem * media;
    NSString       * value;
    
    set = [ NSMutableSet new ];
    
    for( media in array )
    {
        @try
        {
            if( ( value = [ media valueForKeyPath: keypath ] ) && [ value isKindOfClass: [ NSString class ] ] )
            {
                [ set addObject: value ];
            }
        }
        @catch( NSException * e )
        {}
    }
    
    return set.allObjects;
}

- ( void )tableViewSelectionDidChange: ( NSNotification * )notification
{
    NSTableView * tableView;
    BOOL ( ^ artistFilter )( Track * track, NSDictionary * bindings );
    BOOL ( ^ albumFilter  )( Track * track, NSDictionary * bindings );
    
    tableView = ( NSTableView * )( notification.object );
    
    if( [ tableView isKindOfClass: [ NSTableView class ] ] == NO )
    {
        return;
    }
    
    artistFilter = ^ BOOL ( Track * track, NSDictionary * bindings )
    {
        ( void )bindings;
        
        if( self.artistsController.selectedObjects.count == 0 )
        {
            return YES;
        }
        
        for( NSString * artist in self.artistsController.selectedObjects )
        {
            if( [ track.artist isEqualToString: artist ] )
            {
                return YES;
            }
        }
        
        return NO;
    };
    
    albumFilter = ^ BOOL ( Track * track, NSDictionary * bindings )
    {
        ( void )bindings;
        
        if( self.albumsController.selectedObjects.count == 0 )
        {
            return YES;
        }
        
        for( NSString * album in self.albumsController.selectedObjects )
        {
            if( [ track.album isEqualToString: album ] )
            {
                return YES;
            }
        }
        
        return NO;
    };
    
    if( [ tableView.identifier isEqualToString: @"Artists" ] )
    {
        self.tracks = [ self.allTracks filteredArrayUsingPredicate: [ NSPredicate predicateWithBlock: artistFilter ] ];
        self.albums = [ self uniqueValue: @"album"  inArray: self.tracks ];
        
        self.albumsController.selectionIndexes = [ NSIndexSet indexSet ];
        self.tracksController.selectionIndexes = [ NSIndexSet indexSet ];
    }
    else if( [ tableView.identifier isEqualToString: @"Albums" ] )
    {
        self.tracks = [ self.allTracks filteredArrayUsingPredicate: [ NSPredicate predicateWithBlock: artistFilter ] ];
        self.tracks = [ self.tracks    filteredArrayUsingPredicate: [ NSPredicate predicateWithBlock: albumFilter ] ];
        
        self.tracksController.selectionIndexes = [ NSIndexSet indexSet ];
    }
}

@end

File: MediaLibrary/Classes/MainWindowController.swift
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2021 Jean-David Gadina - www.xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

import Cocoa
import iTunesLibrary

@objc class MainWindowController: NSWindowController, NSTableViewDelegate
{
    @IBOutlet private var tracksController:  NSArrayController!
    @IBOutlet private var artistsController: NSArrayController!
    @IBOutlet private var albumsController:  NSArrayController!
    
    @objc private dynamic var allTracks: [ Track ]  = []
    @objc private dynamic var tracks:    [ Track ]  = []
    @objc private dynamic var artists:   [ String ] = []
    @objc private dynamic var albums:    [ String ] = []
    
    override var windowNibName: NSNib.Name?
    {
        "MainWindowController"
    }
    
    override func windowDidLoad()
    {
        super.windowDidLoad()
        
        do
        {
            let library = try ITLibrary( apiVersion: "1.0" )
            
            self.allTracks = library.allMediaItems.map { Track( item: $0 ) }
            self.tracks    = self.allTracks
            self.artists   = self.uniqueValues( for: "artist", in: self.allTracks )
            self.albums    = self.uniqueValues( for: "album",  in: self.allTracks )
            
            self.artistsController.sortDescriptors = [ NSSortDescriptor( key: "description", ascending: true ) ]
            self.albumsController.sortDescriptors  = [ NSSortDescriptor( key: "description", ascending: true ) ]
            self.tracksController.sortDescriptors  =
            [
                NSSortDescriptor( key: "artist", ascending: true ),
                NSSortDescriptor( key: "album",  ascending: true ),
                NSSortDescriptor( key: "title",  ascending: true )
            ]
            
            if let window = self.window
            {
                window.title = "\( window.title ) \( self.allTracks.count ) tracks"
            }
        }
        catch let error
        {
            NSAlert( error: error ).runModal()
        }
    }
    
    private func uniqueValues( for key: String, in array: [ Track ] ) -> [ String ]
    {
        return Array( Set( array.compactMap { $0.value( forKey: key ) as? String } ) )
    }
    
    func tableViewSelectionDidChange( _ notification: Notification )
    {
        guard let tableView = notification.object as? NSTableView else
        {
            return
        }
        
        let artistFilter: ( Track ) -> Bool =
        {
            track in
            
            if self.artistsController.selectedObjects.count == 0
            {
                return true
            }
            
            for artist in self.artistsController.selectedObjects as? [ String ] ?? []
            {
                if track.artist == artist
                {
                    return true
                }
            }
            
            return false
        }
        
        let albumFilter: ( Track ) -> Bool =
        {
            track in
            
            if self.albumsController.selectedObjects.count == 0
            {
                return true
            }
            
            for album in self.albumsController.selectedObjects as? [ String ] ?? []
            {
                if track.album == album
                {
                    return true
                }
            }
            
            return false
        }
        
        self.tracksController.setSelectionIndexes( IndexSet() )
        
        if tableView.identifier == NSUserInterfaceItemIdentifier( rawValue: "Artists" )
        {
            self.tracks = self.allTracks.filter( artistFilter )
            self.albums = self.uniqueValues( for: "album", in: self.tracks )
        }
        else if tableView.identifier == NSUserInterfaceItemIdentifier( rawValue: "Albums" )
        {
            self.tracks = self.allTracks.filter( artistFilter ).filter( albumFilter )
        }
    }
}
